### PR TITLE
Add API for managing volunteering

### PR DESCRIPTION
**Pull request by timyates**
#### Permissions

- Anyone can view organizations
- Anyone can create an organization
- Only users with the `CAN_ADMINISTER_VOLUNTEERING_ORGANIZATIONS` permission can update organizations
- Organizations may be deactivated, but not deleted

#### Data format

```
  {
    "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
    "name": "string",
    "description": "string",
    "website": "string",
    "active": true
  }
```
> [!NOTE]
> `active` defaults to `true` if omitted

#### Endpoints

```
GET /services/volunteer/organization{?includeDeactivated}
POST /services/volunteer/organization
PUT /services/volunteer/organization/{id}
```

#### Permissions

- Anyone can view relationships
- Anyone can create or update a relationship for themselves
- Only users with the `CAN_ADMINISTER_VOLUNTEERING_RELATIONSHIPS` permission can create or update relationships for other people
- Relationships may be deactivated, but not deleted

#### Data format

```
  {
    "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
    "memberId": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
    "organizationId": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
    "startDate": "2024-06-11",
    "endDate": "2024-06-11",
    "active": true
  }
```
> [!NOTE]
> `endDate` is optional and `active` defaults to `true` if omitted

#### Endpoints

```
GET /services/volunteer/relationship{?memberId,organizationId,includeDeactivated}
POST /services/volunteer/relationship
PUT /services/volunteer/relationship/{id}
```

#### Permissions

- Anyone can view events
- Anyone can create, update or delete an event for a relationship that they are the owner of
- Only users with the `CAN_ADMINISTER_VOLUNTEERING_EVENTS` permission can create, update, or delete events for other people's relationships
- Events can be deleted

#### Data format

```
  {
    "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
    "relationshipId": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
    "eventDate": "2024-06-11",
    "hours": 0,
    "notes": "string"
  }
```
> [!NOTE]
> `hours` may not be negative, and `notes` is optional

#### Endpoints

```
GET /services/volunteer/event{?memberId,organizationId,includeDeactivated}
POST /services/volunteer/event
PUT /services/volunteer/event/{id}
DELETE /services/volunteer/event/{id}
```

### CODE REVIEW

**Comment by mkimberlin** [comment]
We should really put required permissions on these APIs. I'm ok with doing a follow up issue for that so this can be merged, if you prefer @mvolkmann @timyates I'll approve and leave that to the two of you.

**Review by mkimberlin — APPROVED** [review]

**Comment by timyates** [comment]
@mvolkmann I've updated the PR description to list the available endpoints and how the permissions work...

Please let me know if I've missed anything, or something could be changed to work better with the front end

**Review by mvolkmann — APPROVED** [review]

File: server/src/main/java/com/objectcomputing/checkins/services/volunteering/VolunteeringServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        if (event.getId() != null) {
+            return update(event);
+        }
+        validateEvent(event, "create");
+        return eventRepo.save(event);
+    }
+
+    @Override
+    public VolunteeringOrganization update(VolunteeringOrganization organization) {
+        // Fail if a certification with the same name already exists (but it's not this one)
```
Comment (vhscom) [inline server/src/main/java/com/objectcomputing/checkins/services/volunteering/VolunteeringServiceImpl.java:106]: Small but here and in the L79 comment, entity changed from `certification` to a volunteering org.

File: server/src/test/java/com/objectcomputing/checkins/services/volunteering/VolunteeringRelationshipControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+        MemberProfile sarah = memberWithoutBoss("sarah");
+        MemberProfile tim = memberWithoutBoss("tim");
+
+        VolunteeringOrganization liftForLife = createVolunteeringOrganization("Lift for Life", "Educate, empower, uplift", "https://www.liftforlifeacademy.org");
+        VolunteeringOrganization foodbank = createVolunteeringOrganization("St. Louis Area Foodbank", "Works with over 600 partners", "https://stlfoodbank.org/find-food/");
+
+        VolunteeringRelationship sarahLiftForLife = createVolunteeringRelationship(sarah.getId(), liftForLife.getId(), LocalDate.now().minusDays(2));
+        VolunteeringRelationship timLiftForLife = createVolunteeringRelationship(tim.getId(), liftForLife.getId(), LocalDate.now());
+        VolunteeringRelationship timFoodbankInactive = createVolunteeringRelationship(tim.getId(), foodbank.getId(), LocalDate.now().minusDays(3), null, false);
+        VolunteeringRelationship sarahFoodbank = createVolunteeringRelationship(sarah.getId(), foodbank.getId(), LocalDate.now().minusDays(10), LocalDate.now());
```
Comment (vhscom) [inline server/src/test/java/com/objectcomputing/checkins/services/volunteering/VolunteeringRelationshipControllerTest.java:228]: The test fixtures make this nice and concise.

**Review by vhscom — APPROVED** [review]
Looks good to me. Left one nice to have.

File: server/src/main/java/com/objectcomputing/checkins/services/volunteering/VolunteeringServiceImpl.java (hunk 1 above)
Comment (timyates) [reply server/src/main/java/com/objectcomputing/checkins/services/volunteering/VolunteeringServiceImpl.java]: Good eyes! Fixed in https://github.com/objectcomputing/check-ins/pull/2490/commits/680d54b73f55a71b1486409e966bdc40944cd020

**Pull request merged**